### PR TITLE
Redirect from www subdomain

### PR DESCRIPTION
**Pull request by sirrus233**
It's considered a best-practice to only have one canonical name for your website. This PR sets up an automatic redirect from `www.twilightimperiumtools.bradleysherman.net` to `twilightimperiumtools.bradleysherman.net`. 

The redirect is implemented with an AWS feature called "Lambda@Edge", which is just a fancy way of saying "a lambda function that runs when the CloudFront endpoint is hit". This lambda function can modify the request before passing it on for further handling.

Lambda@Edge has some pretty severe limitations:
* No custom runtimes (I had a nifty haskell implementation :weary: ), only NodeJS or Python.
* The function must be deployed in the `us-east-1` region
* Typescript is not directly supported***

Since we don't have much testing infrastructure set up - I just deployed this to make sure it works. And it does! But if you think we need any changes, we can easily do that and redeploy.

***It's hard to have a Lambda@Edge function written in typescript, but not impossible. There's a little more information here: https://github.com/aws/aws-cdk/issues/6259

Here's the general deal (took a while to learn all this, so I'm writing it down!):

Lambda's NodeJS runtime doesn't support typescript in general. But if you want to write your functions that way, what you can do is write them in typescript, then transpile to standard JS before deploying to AWS. To avoid a multi-step build/deploy process, you'd ideally do the transpilation as part of your deployment -- and the good news is that CDK supports that! There is a `NodeJSFunction` construct that can accept a typescript file, then at deploy-time uses a docker container to transpile and deploy. Pretty nifty stuff.

Unfortunately, here you hit the Lambda@Edge limitation. The function needs to be in `us-east-1`, and that's not the region of our stack. That means, in order to use the `NodeJSFunction` construct, we'd have to spin up another stack in the `us-east-1` region, define the function resource there, and then use a cross-stack dependency to import the resource from one stack to the other. This can get a little painful.

Enter: the `EdgeFunction` construct! This is a construct that exists solely to make deploying Lambda@Edge functions easier. You can define this construct in a stack for any region (as long as that region is explicitly specified), and it gets created in `us-east-1` for you automagically! I think CloudFormation may be creating a separate regionalized stack under the hood, but I'm not 100% sure about that.

So why isn't the problem solved?? **Because `EdgeFunction` lacks the automatic transpiling of `NodeJSFunction`!!!**. It is truly maddening.

TL;DR Amazon provides constructs that allow for
* Automatic transpiling from typescript -> javascript during deploy
* Automatic cross-region referencing of the function resource

...and you are allowed to have one or the other of these nice things, but not both. RIP :coffin:

### CODE REVIEW

File: infrastructure/bin/infrastructure.ts
Comment (sirrus233) [inline infrastructure/bin/infrastructure.ts:1]: Most of this red diff is just cleaning up the superfluous auto-generated comments from CDK. But there is an important change here: we cannot use the `EdgeFunction` construct without explicitly specifying the region of our stack.

File: infrastructure/eslint.config.mjs
Comment (sirrus233) [inline infrastructure/eslint.config.mjs:1]: Realized we had eslint setup on the `frontend/` directory, but not for `infrastructure/`. Basically just copied over.

File: infrastructure/lambda/redirect/index.mjs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// @ts-check
```
Comment (sirrus233) [inline infrastructure/lambda/redirect/index.mjs:1]: This gives *some* type-checking, as opposed to none. For the reasons why this isn't written in typescript, see the PR description footnote.

More on this: https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html

File: infrastructure/lib/infrastructure-stack.ts [hunk 2]
```diff
@@ -17,6 +19,18 @@ export class InfrastructureStack extends cdk.Stack {
         const bucket = new s3.Bucket(this, "Bucket");
         bucket.grantRead(originAccessIdentity);
 
+        const redirectFunction = new cloudfront.experimental.EdgeFunction(
```
Comment (sirrus233) [inline infrastructure/lib/infrastructure-stack.ts:22]: This is a magic construct. See PR description footnote for details.

File: infrastructure/package.json
Comment (sirrus233) [inline infrastructure/package.json:1]: Auto-reformat - sorry! I just added the eslint dependencies here.

File: infrastructure/lambda/redirect/index.mjs (hunk 1 above)
Comment (chelseaSchmidt) [reply infrastructure/lambda/redirect/index.mjs:1]: TIL, didn't know you could do this, very cool!

File: infrastructure/eslint.config.mjs
Comment (chelseaSchmidt) [reply infrastructure/eslint.config.mjs:1]: Nice 🙏

File: infrastructure/lambda/redirect/index.mjs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+export const handler = async (event) => {
```
Comment (chelseaSchmidt) [inline infrastructure/lambda/redirect/index.mjs:2]: Enjoyed comparing this to the Haskell code you sent, I'm in over my head with Haskell syntax but it helped me understand some of it better!

File: infrastructure/bin/infrastructure.ts
Comment (chelseaSchmidt) [reply infrastructure/bin/infrastructure.ts:1]: ✔️👌

File: infrastructure/package.json
Comment (chelseaSchmidt) [reply infrastructure/package.json:1]: No avoiding it 🥲

**Comment by chelseaSchmidt** [comment]
> It's considered a best-practice to only have one canonical name for your website. This PR sets up an automatic redirect from `www.twilightimperiumtools.bradleysherman.net` to `twilightimperiumtools.bradleysherman.net`.

Awesome 🚀 🎉  tested it out, works like a charm!

> The redirect is implemented with an AWS feature called "Lambda@Edge", which is just a fancy way of saying "a lambda function that runs when the CloudFront endpoint is hit". This lambda function can modify the request before passing it on for further handling.

Checked out the AWS resource created in our stack, very cool!

>  No custom runtimes (I had a nifty haskell implementation 😩 ), only NodeJS or Python.

😭 

> Since we don't have much testing infrastructure set up - I just deployed this to make sure it works. And it does! But if you think we need any changes, we can easily do that and redeploy.

Thank you for deploying! Nice to be able to see it in action 💯 

> Here's the general deal (took a while to learn all this, so I'm writing it down!):

Thank you - one day I'm probably going to need this and will come straight back here for reference 🙏 


> I think CloudFormation may be creating a separate regionalized stack under the hood, but I'm not 100% sure about that.

Totally correct, looked in the console and there's a new CloudFormation stack that only shows up in `us-east-1` now

> So why isn't the problem solved?? **Because `EdgeFunction` lacks the automatic transpiling of `NodeJSFunction`!!!**. It is truly maddening.
> TL;DR Amazon provides constructs that allow for
>  Automatic transpiling from typescript -> javascript during deploy
>  Automatic cross-region referencing of the function resource
> ...and you are allowed to have one or the other of these nice things, but not both. RIP ⚰️

Heartbreaking but this did make me laugh! Awesome sleuthing dude

**Review by chelseaSchmidt — APPROVED** [review]
🚢 🚀 🎉
